contracts/RadiCards.sol

```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
-    );
-  }
-
-  function tokensOf(address _owner) public view returns (uint256[] _tokenIds) {
-    return ownedTokens[_owner];
-  }
-
-  function benefactorsKeys() public view returns (uint256[] _keys) {
-    return benefactorsIndex;
-  }
-
-  function cardsKeys() public view returns (uint256[] _keys) {
-    return cardsIndex;
-  }
-
-  function addBenefactor(uint256 _benefactorIndex, address _ethAddress, string _name, string _website, string _logo)
-  public onlyIfWhitelisted(msg.sender)
-  returns (bool) {
-    require(address(_ethAddress) != address(0), "Invalid address");
-    require(bytes(_name).length != 0, "Invalid name");
-    require(bytes(_website).length != 0, "Invalid name");
-    require(bytes(_logo).length != 0, "Invalid name");
-
-    benefactors[_benefactorIndex] = Benefactor(
-      _ethAddress,
-      _name,
-      _website,
-      _logo
-    );
-    benefactorsIndex.push(_benefactorIndex);
-
-    emit BenefactorAdded(_benefactorIndex);
-    return true;
-  }
 
-  function addCard(uint256 _cardIndex, string _tokenURI, bool _active)
-  public onlyIfWhitelisted(msg.sender)
-  returns (bool) {
-    require(bytes(_tokenURI).length != 0, "Invalid token URI");
-
-    cards[_cardIndex] = CardDesign(
-      _tokenURI,
-      _active
+    event LogClaim(
+            address indexed ephemeralAddress,
+            address indexed sender,
+            uint tokenId,
+            address receiver,
+            uint giftAmount,
+        bool daiDonation
     );
-    cardsIndex.push(_cardIndex);
-
-    emit CardAdded(_cardIndex);
-    return true;
-  }
-
-  function setTokenBaseURI(string _newBaseURI) external onlyIfWhitelisted(msg.sender) {
-    require(bytes(_newBaseURI).length != 0, "Base URI invalid");
 
-    tokenBaseURI = _newBaseURI;
-  }
+    event LogCancel(
+        address indexed ephemeralAddress,
+        address indexed sender,
+        uint tokenId
+    );
 
-  function setMinContribution(uint256 _minContribution) external onlyIfWhitelisted(msg.sender) {
-    require(_minContribution > 0, "Invalid minimum contribution");
 
-    minContribution = _minContribution;
-  }
+    event BenefactorAdded(
+        uint256 indexed _benefactorIndex
+    );
 
-  function setActive(uint256 _cardIndex, bool _active) external onlyIfWhitelisted(msg.sender) {
-    require(bytes(cards[_cardIndex].tokenURI).length != 0, "Must specify existing card");
+    event CardAdded(
+        uint256 indexed _cardIndex
+    );
 
-    cards[_cardIndex].active = _active;
-  }
+    constructor () public ERC721Token("RadiCards", "RADI") {
+        addAddressToWhitelist(msg.sender);
+    }
+
+    function gift(address to, uint256 _benefactorIndex, uint256 _cardIndex, string _message, uint256 _donationAmount, uint256 _giftAmount, bool _claimableLink) payable public returns (bool) {
+        require(to != address(0), "Must be a valid address");
+        require(benefactors[_benefactorIndex].ethAddress != address(0), "Must specify existing benefactor");
```

Comment:
so we need to pick a benefactor even if we aren't sending anything to them? We could remove the `require` as the UI or whatever may just want to pass `address(0)` when not relevant?

Reply:
I agree with this logic. will implement